aptos-move/aptos-vm/src/aptos_vm.rs

```diff
@@ -197,10 +197,39 @@ pub(crate) fn get_or_vm_startup_failure<'a, T>(
     })
 }
 
+/// Checks if a given transaction is a governance proposal by checking if it has one of the
+/// approved execution hashes.
+fn is_governance_proposal<R, E>(
+    resolver: &R,
+    txn: &SignedTransaction,
+    txn_metadata: &TransactionMetadata,
+) -> bool
+where
+    R: MoveResolver<E>,
+    E: Debug,
+{
+    match txn.payload() {
+        TransactionPayload::Script(_script) => {
+            match resolver.get_resource(&CORE_CODE_ADDRESS, &ApprovedExecutionHashes::struct_tag())
+            {
+                Ok(Some(bytes)) => match bcs::from_bytes::<ApprovedExecutionHashes>(&bytes) {
+                    Ok(approved_execution_hashes) => approved_execution_hashes
+                        .entries
+                        .iter()
+                        .any(|(_, hash)| hash == &txn_metadata.script_hash),
+                    Err(_) => false,
+                },
+                _ => false,
```

Comment:
nit: maybe you cam map and flatten to avoid these cases which all go to false

Reply:
It's probably a matter of personal taste, but I've always found combinators reducing readability when there are nested control flows, as you have to keep track of both the happy and error paths until the end.